src/librbd/crypto/luks/FormatLUKSRequest.cc

```diff
@@ -0,0 +1,122 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "FormatLUKSRequest.h"
+#include <errno.h>
+
+#include "common/dout.h"
+#include "librbd/Utils.h"
+#include "librbd/api/Io.h"
+#include "librbd/crypto/luks/Utils.h"
+#include "librbd/io/AioCompletion.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::crypto::luks::FormatLUKSRequest: " \
+                           << this << " " << __func__
+
+namespace librbd {
+namespace crypto {
+namespace luks {
+
+using librbd::util::create_context_callback;
+
+template <typename I>
+FormatLUKSRequest<I>::FormatLUKSRequest(
+        I* image_ctx, const char* passphrase, size_t passphrase_size,
+        Context* on_finish) : m_image_ctx(image_ctx), m_on_finish(on_finish),
+                              m_fd(-1), m_cd(nullptr),
+                              m_passphrase_size(passphrase_size) {
+  m_passphrase = new char[passphrase_size];
+  memcpy(m_passphrase, passphrase, passphrase_size);
+}
+
+template <typename I>
+void FormatLUKSRequest<I>::send() {
+  auto [m_fd, path] = create_anonymous_file("LUKSHeader");
+  if (m_fd == -1) {
+    lderr(m_image_ctx->cct) << "error creating anonymous file: " << errno
+                            << dendl;
+    complete(errno);
+    return;
+  }
+
+  // init libcryptsetup handle
+  auto r = crypt_init(&m_cd, path.c_str());
+  if (r < 0) {
+    lderr(m_image_ctx->cct) << "crypt_init failed: " << r << dendl;
+    complete(r);
+    return;
+  }
+
+  // format (in-memory)
+  r = crypt_format(
+          m_cd, CRYPT_LUKS2, "aes", "xts-plain64", NULL, NULL, 64, NULL);
+  if (r < 0) {
+    lderr(m_image_ctx->cct) << "crypt_format failed: " << r << dendl;
+    complete(r);
+    return;
+  }
+
+  // create keyslot (volume key encrypted by passphrase)
+  r = crypt_keyslot_add_by_volume_key(m_cd, CRYPT_ANY_SLOT, NULL, 0,
+                                      m_passphrase, m_passphrase_size);
+  if (r < 0) {
+    lderr(m_image_ctx->cct) << "crypt_keyslot_add_by_volume_key failed: "
+                            << r << dendl;
+    complete(r);
+    return;
+  }
+
+  // construct header data in a bufferlist
+  // TODO: switch to crypt_get_metadata_size when it becomes available
+  ceph::bufferlist bl;
+  auto keyslots_end_offset =
+          crypt_get_data_offset(m_cd) * crypt_get_sector_size(m_cd);
+  auto bytes_read = bl.read_fd(m_fd, keyslots_end_offset);
+  if (bytes_read < 0) {
+    lderr(m_image_ctx->cct) << "error reading header from memory: "
+                            << bytes_read << dendl;
+    complete(bytes_read);
+    return;
+  }
+
+  // write header to offset 0 of the image
+  auto ctx = create_context_callback<
+          FormatLUKSRequest<I>,
+          &FormatLUKSRequest<I>::handle_write_header>(this);
+  auto aio_comp = io::AioCompletion::create_and_start(
+          ctx, util::get_image_ctx(m_image_ctx), io::AIO_TYPE_WRITE);
+  api::Io<I>::aio_write(
```

Comment:
Nit: create a `io::ImageDispatchSpec` at the `CRYPTO` layer and send it.

Reply:
This should happen before the crypto layer is actually loaded.

Reply:
Ack -- no worries, it will still skip past the (non-existent) layer and keep the IO dispatching consistent.

Reply:
First, do you actually want to support a case where format is called while the crypto layer is already loaded?
Second, assuming you do, starting with the crypto layer is currently not enough since:
1. You still don't skip the object crypto layer.
2. The image crypto layer as I implemented it in #37935 is a total pass-through for any IO, and only comes to play in the CORE layer which will go again through all of the registered layers.